# Regular Expressions

In [1]:
import pandas as pd
import numpy as np
import re

### 1)
   Write a function named is_vowel. It should accept a string as input and use a regular expression to determine if the passed string is a vowel. While not explicity mentioned in the lesson, you can treat the result of re.search as a boolean value that indicates whether or not the regular expression matches the given string.

In [319]:
def is_vowel():
    subject = input('Enter a character: ')
    regexp = re.compile(r'^[aeiou]$', re.IGNORECASE)
    if regexp.search(subject):
        return True
    else:
        return False

In [320]:
is_vowel()

Enter a character:  A


True

### 2)
Write a function named is_valid_username that accepts a string as input. A valid username starts with a lowercase letter, and only consists of lowercase letters, numbers, or the _ character. It should also be no longer than 32 characters. The function should return either True or False depending on whether the passed string is a valid username.

In [321]:
def is_valid_username():
    subject = input('Enter a username')
    regexp = r'\b[a-z][a-z0-9_]{0,31}$'
    if re.match(regexp, subject):
        return True
    else:
        return False

>>> is_valid_username('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa')
False
>>> is_valid_username('codeup')
True
>>> is_valid_username('Codeup')
False
>>> is_valid_username('codeup123')
True
>>> is_valid_username('1codeup')
False


In [322]:
is_valid_username()

Enter a username aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa


False

In [323]:
is_valid_username()

Enter a username codeup


True

In [324]:
is_valid_username()

Enter a username Codeup


False

In [325]:
is_valid_username()

Enter a username codeup123


True

In [326]:
is_valid_username()

Enter a username 1codeup


False

In [327]:
is_valid_username()

Enter a username theo123


True

### 3)
Write a regular expression to capture phone numbers. It should match all of the following:

In [328]:
def is_phone_number():
    subject = input('Enter a phone number')
    regexp = r'\+?1?\s?(\(?\d{3}?\)?)?[-\.\s]?\d{3}[-\.\s]\d{4}$'
    if re.match(regexp, subject):
        return True
    else:
        return False

In [329]:
is_phone_number()

Enter a phone number (210) 867 5309


True

In [330]:
is_phone_number()

Enter a phone number +1 210.867.5309


True

In [331]:
is_phone_number()

Enter a phone number 867-5309


True

In [332]:
is_phone_number()

Enter a phone number 210-867-5309


True

(210) 867 5309
+1 210.867.5309
867-5309
210-867-5309


### 4) 
Use regular expressions to convert the dates below to the standardized year-month-day format.

In [333]:
def convert_date():
    user_input = input("Enter a date (MM/DD/YY): ")
    match = re.match(r'(\d{2})/(\d{2})/(\d{2})', user_input)
    month, day, year = match.groups()
    return f"20{year}-{month}-{day}"

In [334]:
convert_date()

Enter a date (MM/DD/YY):  02/04/19


'2019-02-04'

In [335]:
convert_date()

Enter a date (MM/DD/YY):  02/05/19


'2019-02-05'

In [336]:
convert_date()

Enter a date (MM/DD/YY):  02/06/19


'2019-02-06'

In [337]:
convert_date()

Enter a date (MM/DD/YY):  02/07/19


'2019-02-07'

In [338]:
convert_date()

Enter a date (MM/DD/YY):  02/08/19


'2019-02-08'

In [339]:
convert_date()

Enter a date (MM/DD/YY):  02/09/19


'2019-02-09'

In [340]:
convert_date()

Enter a date (MM/DD/YY):  02/10/19


'2019-02-10'

02/04/19
02/05/19
02/06/19
02/07/19
02/08/19
02/09/19
02/10/19


### 5) 
Write a regex to extract the various parts of these logfile lines:

GET /api/v1/sales?page=86 [16/Apr/2019:193452+0000] HTTP/1.1 {200} 510348 "python-requests/2.21.0" 97.105.19.58  
POST /users_accounts/file-upload [16/Apr/2019:193452+0000] HTTP/1.1 {201} 42 "User-Agent: Mozilla/5.0 (X11; Fedora; Fedora; Linux x86_64) AppleWebKit/537.36   (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36" 97.105.19.58  
GET /api/v1/items?page=3 [16/Apr/2019:193453+0000] HTTP/1.1 {429} 3561 "python-requests/2.21.0" 97.105.19.58  

In [343]:
log_lines = [
    'GET /api/v1/sales?page=86 [16/Apr/2019:193452+0000] HTTP/1.1 {200} 510348 "python-requests/2.21.0" 97.105.19.58',
    'POST /users_accounts/file-upload [16/Apr/2019:193452+0000] HTTP/1.1 {201} 42 "User-Agent: Mozilla/5.0 (X11; Fedora; Fedora; Linux x86_64) AppleWebKit/537.36   (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36" 97.105.19.58',
    'GET /api/v1/items?page=3 [16/Apr/2019:193453+0000] HTTP/1.1 {429} 3561 "python-requests/2.21.0" 97.105.19.58'
]

### Method

In [341]:
for method in log_lines:
    match = re.search(r'^(GET|POST)', method)
    if match:
        print('Match:', match.group(1))
    else:
        print('No match')

Match: GET
Match: POST
Match: GET


### log

In [342]:
for log in log_lines:
    match = re.search(r' /([a-zA-Z0-9_/?=-]+)', log)
    if match:
        print('Match:', match.group(1))
    else:
        print('No match')

Match: api/v1/sales?page=86
Match: users_accounts/file-upload
Match: api/v1/items?page=3


### Timestamp

In [344]:
for timestamp in log_lines:
    match = re.search(r'\[(\d{2}/[A-Za-z]+/\d{4}:\d{6}\+\d{4})', timestamp)
    if match:
        print('Match:', match.group(1))
    else:
        print('No match')

Match: 16/Apr/2019:193452+0000
Match: 16/Apr/2019:193452+0000
Match: 16/Apr/2019:193453+0000


### httpversion

In [346]:
for httpversion in log_lines:
    match = re.search(r'HTTP/([0-9.]+)', httpversion)
    if match:
        print('Match:', match.group(1))
    else:
        print('No match')

Match: 1.1
Match: 1.1
Match: 1.1


### Httpcode

In [348]:
for httpcode in log_lines:
    match = re.search(r'{(\d+)}', httpcode)
    if match:
        print('Match:', match.group(1))
    else:
        print('No match')

Match: 200
Match: 201
Match: 429


### Bytes

In [350]:
for bytes in log_lines:
    match = re.search(r'(\d+) "(.*?)"', bytes)
    if match:
        print('Match:', match.group(1))
    else:
        print('No match')

Match: 510348
Match: 42
Match: 3561


### Useragent

In [352]:
for useragent in log_lines:
    match = re.search(r'"(.*?)"', useragent)
    if match:
        print('Match:', match.group(1))
    else:
        print('No match')

Match: python-requests/2.21.0
Match: User-Agent: Mozilla/5.0 (X11; Fedora; Fedora; Linux x86_64) AppleWebKit/537.36   (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36
Match: python-requests/2.21.0


### IP

In [354]:
for ip in log_lines:
    match = re.search(r'(\d+\.\d+\.\d+\.\d+)', ip)
    if match:
        print('Match:', match.group(1))
    else:
        print('No match')

Match: 97.105.19.58
Match: 73.0.3683.86
Match: 97.105.19.58


## Different method

In [345]:
logfile_re = r'''
^(?P<method>GET|POST)
\s+
(?P<path>.*?)
\s+
\[(?P<timestamp>.*?)\]
\s+
(?P<http_version>.*?)
\s+
\{(?P<status>\d+)\}
\s+
(?P<bytes>\d+)
\s+
"(?P<user_agent>.*)"
\s+
(?P<ip>.*)$
'''

lines = pd.Series([
    'GET /api/v1/sales?page=86 [16/Apr/2019:193452+0000] HTTP/1.1 {200} 510348 "python-requests/2.21.0" 97.105.19.58',
    'POST /users_accounts/file-upload [16/Apr/2019:193452+0000] HTTP/1.1 {201} 42 "User-Agent: Mozilla/5.0 (X11; Fedora; Fedora; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36" 97.105.19.58',
    'GET /api/v1/items?page=3 [16/Apr/2019:193453+0000] HTTP/1.1 {429} 3561 "python-requests/2.21.0" 97.105.19.58',
])
lines.str.extract(logfile_re, re.VERBOSE)

method                         path                timestamp http_version  \
0    GET        /api/v1/sales?page=86  16/Apr/2019:193452+0000     HTTP/1.1   
1   POST  /users_accounts/file-upload  16/Apr/2019:193452+0000     HTTP/1.1   
2    GET         /api/v1/items?page=3  16/Apr/2019:193453+0000     HTTP/1.1   

  status   bytes                                         user_agent  \
0    200  510348                             python-requests/2.21.0   
1    201      42  User-Agent: Mozilla/5.0 (X11; Fedora; Fedora; ...   
2    429    3561                             python-requests/2.21.0   

             ip  
0  97.105.19.58  
1  97.105.19.58  
2  97.105.19.58